src/sinks/util/sink.rs

```diff
@@ -0,0 +1,1193 @@
+            batch,
+            settings,
+            acker,
+            DefaultExecutor::current(),
+        )
+    }
+}
+
+impl<B, S, K, Request, E> PartitionBatchSink<B, S, K, Request, E>
+where
+    B: Batch<Output = Request>,
+    B::Input: Partition<K>,
+    K: Hash + Eq + Clone + Send + 'static,
+    S: Service<Request>,
+    S::Future: Send + 'static,
+    S::Error: Into<crate::Error> + Send + 'static,
+    S::Response: fmt::Debug,
+    E: Executor,
+{
+    pub fn with_executor(
+        service: S,
+        batch: B,
+        settings: BatchSettings,
+        acker: Acker,
+        exec: E,
+    ) -> Self {
+        let service = ServiceSink::new(service, acker);
+
+        Self {
+            batch,
+            service,
+            exec,
+            partitions: HashMap::new(),
+            settings,
+            closing: false,
+            sending: VecDeque::new(),
+            lingers: FuturesUnordered::new(),
+            linger_handles: HashMap::new(),
+        }
+    }
+
+    fn set_linger(&mut self, partition: K) {
+        let (tx, rx) = oneshot::channel();
+        let partition_clone = partition.clone();
+
+        let deadline = Instant::now() + self.settings.timeout;
+        let delay = Delay::new(deadline)
+            .map(move |_| LingerState::Elapsed(partition_clone))
+            .map_err(|_| ());
+
+        let cancel = rx.map(|_| LingerState::Canceled).map_err(|_| ());
+
+        let fut = cancel
+            .select2(delay)
+            .map(|state| match state {
+                Either::A((state, _)) => state,
+                Either::B((state, _)) => state,
+            })
+            .map_err(|_| ());
+
+        self.linger_handles.insert(partition, tx);
+        self.lingers.push(Box::new(fut));
+    }
+
+    fn poll_send(&mut self, batch: B) -> Poll<(), crate::Error> {
+        if let Async::NotReady = self.service.poll_ready()? {
+            self.sending.push_front(batch);
+            Ok(Async::NotReady)
+        } else {
+            let batch_size = batch.num_items();
+            let batch = batch.finish();
+            let fut = self.service.call(batch, batch_size);
+
+            self.exec.spawn(fut).unwrap();
+
+            self.service.poll_complete()
+        }
+    }
+}
+
+impl<B, S, K, Request, E> Sink for PartitionBatchSink<B, S, K, Request, E>
+where
+    B: Batch<Output = Request>,
+    B::Input: Partition<K>,
+    K: Hash + Eq + Clone + Send + 'static,
+    S: Service<Request>,
+    S::Future: Send + 'static,
+    S::Error: Into<crate::Error> + Send + 'static,
+    S::Response: fmt::Debug,
+    E: Executor,
+{
+    type SinkItem = B::Input;
+    type SinkError = crate::Error;
+
+    fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        // Apply back pressure if we are buffering more than
+        // 5 batches, this should only happen if the inner sink
+        // is apply back pressure.
+        if self.sending.len() > 5 {
```

Comment:
Wouldn't this cause backpressure very quickly if we have data with more than 5 partitions? I'm not quite sure what the goal is here.

Reply:
This was set before, I was thinking about this but I actually don't think its even possible to go beyond 1 here since the semantics of service. I will take a look tomorrow.

Reply:
Ok, so I looked into this more the reason I added it originally was to work around multiple batches being ready and the service applying back pressure. Since, we can fill multiple batches at once we want to buffer the complete ones until we can send them so we can continue to fill. Since, in theory the service should only be applying back pressure for a short amount of time this should only live here for a short period.

I also want to avoid refactoring this piece since it seems to be working correctly and we have not noticed issues. I suspect a lot of this code wil go away in the future and replaced by nicer async/await so I want to avoid investing too much. A lot of these types of work arounds are for the sink trait.